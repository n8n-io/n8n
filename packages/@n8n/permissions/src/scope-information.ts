import { API_KEY_RESOURCES, RESOURCES } from './constants.ee';
import type { ApiKeyScope, Scope, ScopeInformation } from './types.ee';

function buildResourceScopes() {
	const resourceScopes = Object.entries(RESOURCES).flatMap(([resource, operations]) => [
		...operations.map((op) => `${resource}:${op}` as const),
		`${resource}:*` as const,
	]) as Scope[];

	resourceScopes.push('*' as const); // Global wildcard
	return resourceScopes;
}

function buildApiKeyScopes() {
	const apiKeyScopes = Object.entries(API_KEY_RESOURCES).flatMap(([resource, operations]) => [
		...operations.map((op) => `${resource}:${op}` as const),
	]) as ApiKeyScope[];

	return new Set(apiKeyScopes);
}

export const ALL_SCOPES = buildResourceScopes();

export const ALL_API_KEY_SCOPES = buildApiKeyScopes();

export const scopeInformation: Partial<Record<Scope, ScopeInformation>> = {
	'annotationTag:create': {
		displayName: 'Create Annotation Tag',
		description: 'Allows creating new annotation tags.',
	},
};
