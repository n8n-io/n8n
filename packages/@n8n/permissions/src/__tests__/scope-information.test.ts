import { ALL_SCOPES } from '@/scope-information';

describe('Scope Information', () => {
	it('ensure scopes are defined correctly', () => {
		expect(ALL_SCOPES).toMatchSnapshot();
	});
});
