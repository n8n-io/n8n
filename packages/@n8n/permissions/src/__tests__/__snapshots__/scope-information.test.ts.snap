// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scope Information ensure scopes are defined correctly 1`] = `
[
  "annotationTag:create",
  "annotationTag:read",
  "annotationTag:update",
  "annotationTag:delete",
  "annotationTag:list",
  "annotationTag:*",
  "auditLogs:manage",
  "auditLogs:*",
  "banner:dismiss",
  "banner:*",
  "community:register",
  "community:*",
  "communityPackage:install",
  "communityPackage:uninstall",
  "communityPackage:update",
  "communityPackage:list",
  "communityPackage:manage",
  "communityPackage:*",
  "credential:share",
  "credential:move",
  "credential:create",
  "credential:read",
  "credential:update",
  "credential:delete",
  "credential:list",
  "credential:*",
  "externalSecretsProvider:sync",
  "externalSecretsProvider:create",
  "externalSecretsProvider:read",
  "externalSecretsProvider:update",
  "externalSecretsProvider:delete",
  "externalSecretsProvider:list",
  "externalSecretsProvider:*",
  "externalSecret:list",
  "externalSecret:use",
  "externalSecret:*",
  "eventBusDestination:test",
  "eventBusDestination:create",
  "eventBusDestination:read",
  "eventBusDestination:update",
  "eventBusDestination:delete",
  "eventBusDestination:list",
  "eventBusDestination:*",
  "ldap:sync",
  "ldap:manage",
  "ldap:*",
  "license:manage",
  "license:*",
  "logStreaming:manage",
  "logStreaming:*",
  "orchestration:read",
  "orchestration:list",
  "orchestration:*",
  "project:create",
  "project:read",
  "project:update",
  "project:delete",
  "project:list",
  "project:*",
  "saml:manage",
  "saml:*",
  "securityAudit:generate",
  "securityAudit:*",
  "sourceControl:pull",
  "sourceControl:push",
  "sourceControl:manage",
  "sourceControl:*",
  "tag:create",
  "tag:read",
  "tag:update",
  "tag:delete",
  "tag:list",
  "tag:*",
  "user:resetPassword",
  "user:changeRole",
  "user:enforceMfa",
  "user:create",
  "user:read",
  "user:update",
  "user:delete",
  "user:list",
  "user:*",
  "variable:create",
  "variable:read",
  "variable:update",
  "variable:delete",
  "variable:list",
  "variable:*",
  "workersView:manage",
  "workersView:*",
  "workflow:share",
  "workflow:execute",
  "workflow:move",
  "workflow:activate",
  "workflow:deactivate",
  "workflow:create",
  "workflow:read",
  "workflow:update",
  "workflow:delete",
  "workflow:list",
  "workflow:*",
  "folder:create",
  "folder:read",
  "folder:update",
  "folder:delete",
  "folder:list",
  "folder:move",
  "folder:*",
  "insights:list",
  "insights:*",
  "oidc:manage",
  "oidc:*",
  "dataStore:create",
  "dataStore:read",
  "dataStore:update",
  "dataStore:delete",
  "dataStore:list",
  "dataStore:readRow",
  "dataStore:writeRow",
  "dataStore:listProject",
  "dataStore:*",
  "execution:delete",
  "execution:read",
  "execution:list",
  "execution:get",
  "execution:*",
  "workflowTags:update",
  "workflowTags:list",
  "workflowTags:*",
  "*",
]
`;
